model/log.go

```diff
@@ -1,198 +1,140 @@
+func DownloadLogLines(ctx context.Context, buildID string, testID string) (chan *LogLineItem, error) {
+	buildKeys, err := getBuildKeys(ctx, buildID)
 	if err != nil {
-		return nil, errors.Wrap(err, "finding global logs during test")
+		return nil, errors.Wrapf(err, "getting keys for build '%s'", buildID)
 	}
-	testLogs := findLogsInWindow(bson.M{"build_id": test.BuildId, "test_id": test.Id}, []string{"seq"}, nil, nil)
-	return MergeLogChannels(testLogs, globalLogs), nil
-}
 
-func findGlobalLogsDuringTest(test *Test) (chan *LogLineItem, error) {
-	db, closeSession := db.DB()
-	defer closeSession()
-
-	var globalSeqFirst, globalSeqLast *int
+	if len(buildKeys) == 0 {
+		return nil, errors.Errorf("no keys found for build '%s", buildID)
+	}
 
-	minTime, maxTime, err := test.GetExecutionWindow()
+	buildChunks, testChunks, err := parseLogChunks(buildKeys)
 	if err != nil {
-		return nil, errors.Wrap(err, "getting execution window")
+		return nil, errors.Wrapf(err, "parsing log chunks from keys for build '%s'", buildID)
 	}
+	testChunks = filterLogChunksByTestID(testChunks, testID)
 
-	// Find the first global log entry before this test started.
-	// This may not actually contain any global log lines during the test run, if the entry returned
-	// by this query comes from after the *next* test stared.
-	firstGlobalLog := &Log{}
-	err = db.C("logs").Find(bson.M{"build_id": test.BuildId, "test_id": nil, "started": bson.M{"$lt": minTime}}).Sort("-seq").Limit(1).One(firstGlobalLog)
+	testIDs, err := parseTestIDs(buildKeys)
 	if err != nil {
-		if err != mgo.ErrNotFound {
-			return nil, err
-		}
-		// There are no global entries after this test started.
-		globalSeqFirst = nil
-	} else {
-		globalSeqFirst = &firstGlobalLog.Seq
+		return nil, errors.Wrapf(err, "parsing test IDs from keys for build '%s'", buildID)
 	}
-
-	lastGlobalLog := &Log{}
-
-	if maxTime != nil {
-		// Find the last global log entry that covers this test. This may return a global log entry
-		// that started before the test itself.
-		err = db.C("logs").Find(bson.M{"build_id": test.BuildId, "test_id": nil, "started": bson.M{"$lt": maxTime}}).Sort("-seq").Limit(1).One(lastGlobalLog)
-		if err != nil {
-			if err != mgo.ErrNotFound {
-				return nil, err
-			}
-			globalSeqLast = nil
-		} else {
-			globalSeqLast = &lastGlobalLog.Seq
-		}
+	tr, err := testExecutionWindow(testIDs, testID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "getting execution window for test '%s'", testID)
 	}
 
-	var globalLogsSeq bson.M
-	if globalSeqFirst == nil {
-		globalLogsSeq = bson.M{"$gte": test.Seq}
+	// Tests should never be filtered by a time range other than AllTime
+	// since we always want to capture all the lines of either a single
+	// test or all tests.
+	return NewMergingIterator(NewBatchedLogIterator(testChunks, 4, AllTime), NewBatchedLogIterator(buildChunks, 4, tr)).Stream(ctx), nil
+}
+
+func (item *LogLineItem) Color() string {
```

Comment:
Documentation.

Reply:
done
When I figured out what it was doing I realized the function wasn't named well so I renamed it.